File: HackerRank/30 days of code/day 18, queues and stacks/day_18_queues_and_stacks.java
import java.io.*;
import java.util.*;

public class Solution {
    private Stack<Character> myStack;
    private PriorityQueue<Character> myQueue;

    Solution() {
        myStack = new Stack<Character>();
        myQueue = new PriorityQueue<Character>();
    }

    void pushCharacter(char ch) {
        myStack.push(ch);
    }

    void enqueueCharacter(char ch) {
        myQueue.add(ch);
    }

    char popCharacter() {
        return myStack.pop().charValue();
    }

    char dequeueCharacter() {
        return myQueue.poll().charValue();
    }

    boolean empty() {
        return myStack.empty();
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        String input = scan.nextLine();
        scan.close();

        // Convert input String to an array of characters:
        char[] s = input.toCharArray();

        // Create a Solution object:
        Solution p = new Solution();

        // Enqueue/Push all chars to their respective data structures:
        for (char c : s) {
            p.pushCharacter(c);
            p.enqueueCharacter(c);
        }

        // Pop/Dequeue the chars at the head of both data structures and compare them:
        boolean isPalindrome = true;
        for (int i = 0; i < s.length/2; i++) {
            if (p.popCharacter() != p.dequeueCharacter()) {
                isPalindrome = false;                
                break;
            }
        }

        //Finally, print whether string s is palindrome or not.
        System.out.println( "The word, " + input + ", is " 
                           + ( (!isPalindrome) ? "not a palindrome." : "a palindrome." ) );
    }
}

File: HackerRank/30 days of code/day 18, queues and stacks/day_18_queues_and_stacks.js
process.stdin.resume();
process.stdin.setEncoding('ascii');

var input_stdin = "";
var input_stdin_array = "";
var input_currentline = 0;

process.stdin.on('data', function (data) {
    input_stdin += data;
});

process.stdin.on('end', function () {
    input_stdin_array = input_stdin.split("\n");
    main();    
});

function readLine() {
    return input_stdin_array[input_currentline++];
}

function Solution() {
    this.myQueue = []
    this.myStack = []  

    this.pushCharacter = ch => this.myStack.push(ch)

    this.enqueueCharacter = ch => this.myQueue.push(ch)

    this.popCharacter = () => {return this.myStack.pop()}

    this.dequeueCharacter = () => {return this.myQueue.shift()}
}

// HackerRank immutable code below
function main(){
    // read the string s
    var s=readLine();
    var len=s.length;
    // create the Solution class object p
    var obj=new Solution();
    //push/enqueue all the characters of string s to stack
    for(var i=0;i<len;i++){
        obj.pushCharacter(s.charAt(i));
        obj.enqueueCharacter(s.charAt(i));
    }
  
    var isPalindrome=true;
    /*
    pop the top character from stack
    dequeue the first character from queue
    compare both the characters*/

    for(var i=0;i<len/2;i++){
        if(obj.popCharacter()!=obj.dequeueCharacter()){
            isPalindrome=false;
            break;
        }
    }
    //finally print whether string s is palindrome or not

    if(isPalindrome)
        console.log("The word, "+s+", is a palindrome.");
    else
        console.log("The word, "+s+", is not a palindrome.");
}